vql/tools/sftp_upload.go

```diff
@@ -0,0 +1,202 @@
+	if err != nil {
+		scope.Log("upload_sftp: %s", err.Error())
+		return vfilter.Null{}
+	}
+
+	err = vql_subsystem.CheckFilesystemAccess(scope, arg.Accessor)
+	if err != nil {
+		scope.Log("upload_SFTP: %s", err)
+		return vfilter.Null{}
+	}
+
+	accessor, err := glob.GetAccessor(arg.Accessor, scope)
+	if err != nil {
+		scope.Log("upload_SFTP: %v", err)
+		return vfilter.Null{}
+	}
+
+	file, err := accessor.Open(arg.File)
+	if err != nil {
+		scope.Log("upload_SFTP: Unable to open %s: %s",
+			arg.File, err.Error())
+		return &vfilter.Null{}
+	}
+	defer file.Close()
+
+	if arg.Name == "" {
+		arg.Name = arg.File
+	}
+
+	stat, err := file.Stat()
+	if err != nil {
+		scope.Log("upload_SFTP: Unable to stat %s: %v",
+			arg.File, err)
+	} else if !stat.IsDir() {
+		// Abort uploading when the scope is destroyed.
+		sub_ctx, cancel := context.WithCancel(ctx)
+		scope.AddDestructor(func() {
+			cancel()
+		})
+
+		upload_response, err := upload_SFTP(
+			sub_ctx, scope, file,
+			strings.TrimSpace(arg.User),
+			strings.TrimSpace(arg.Path),
+			strings.TrimSpace(arg.Name),
+			strings.TrimSpace(arg.PrivateKey),
+			strings.TrimSpace(arg.Endpoint),
+			strings.TrimSpace(arg.HostKey))
+		if err != nil {
+			scope.Log("upload_SFTP: %v", err)
+			// Relay the error in the UploadResponse
+			return upload_response
+		}
+		return upload_response
+	}
+
+	return vfilter.Null{}
+}
+
+func keyString(k ssh.PublicKey) string {
+	return k.Type() + " " + base64.StdEncoding.EncodeToString(k.Marshal())
+}
+
+func hostkeycallback(trustedkey string) ssh.HostKeyCallback {
+	return func(_ string, _ net.Addr, k ssh.PublicKey) error {
+		ks := keyString(k)
+		if trustedkey != ks {
+			return fmt.Errorf("SSH-key verification: expected %q but got %q", trustedkey, ks)
+		}
+		return nil
+	}
+}
+
+func upload_SFTP(ctx context.Context, scope *vfilter.Scope,
+	reader io.Reader,
+	user, path, name string,
+	privateKey string, endpoint string, hostkey string) (
+	*api.UploadResponse, error) {
+
+	scope.Log("upload_SFTP: Uploading %v to %v", name, endpoint)
+	signer, err := ssh.ParsePrivateKey([]byte(strings.TrimSuffix(privateKey, "\n")))
+	if err != nil {
+		return &api.UploadResponse{
+			Error: err.Error(),
+		}, err
+	}
+
+	var clientConfig *ssh.ClientConfig
+	if hostkey == "" {
+		clientConfig = &ssh.ClientConfig{
+			User: user,
+			Auth: []ssh.AuthMethod{
+				ssh.PublicKeys(signer),
+			},
+			HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		}
+	} else {
+		clientConfig = &ssh.ClientConfig{
+			User: user,
+			Auth: []ssh.AuthMethod{
+				ssh.PublicKeys(signer),
+			},
+			HostKeyCallback: hostkeycallback(hostkey),
+		}
+	}
+
+	conn, err := ssh.Dial("tcp", endpoint, clientConfig)
+	if err != nil {
+		return &api.UploadResponse{
+			Error: err.Error(),
+		}, err
+	}
+	defer conn.Close()
+	client, err := sftp.NewClient(conn)
+	if err != nil {
+		return &api.UploadResponse{
+			Error: err.Error(),
+		}, err
+	}
+	defer client.Close()
```

Comment:
Maybe cache the client in the query scope? Getting the client in the first place is quite an expensive process of connection and key verification but once it is there you should be able to reuse  it very cheaply.

This way if this VQL plugin is used many times in the same query (e.g. in uploading an entire directory) then it will be very quick.